compiler/rustc_const_eval/src/interpret/terminator.rs

```diff
@@ -115,6 +88,45 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 }
             }
 
+            TailCall { ref func, ref args, fn_span: _ } => {
+                let old_frame_idx = self.frame_idx();
+
+                let (fn_val, args, fn_sig, fn_abi, with_caller_location) =
+                    self.prepare_fn_for_call(terminator, func, args)?;
+
+                // This is the "canonical" implementation of tails calls,
+                // a pop of the current stack frame, followed by a normal call
+                // which pushes a new stack frame, with the return address from
+                // the popped stack frame.
+                //
+                // Note that we can't use `pop_stack_frame` as it "executes"
+                // the goto to the return block, but we don't want to,
+                // only the tail called function should return to the current
+                // return block.
```

Comment:
`pop_stack_frame` does a bunch of other stuff as well though that seems relevant, such as calling `after_stack_pop`.

What if the old function has return type `bool`, but the callee has return type `u8` and returns `2` -- is that UB? I would think so but currently I think that would be missed. (Even if the static checks ensure this can't usually happen, we can always transmute a `fn() -> bool` to `fn() -> u8` to circumvent those checks.)

Reply:
> pop_stack_frame does a bunch of other stuff as well though that seems relevant, such as calling after_stack_pop.

In particular this is relevant for the aliasing model, where protectors are active until the stack frame pops.

This is actually an interesting question for the aliasing model... should protector remain active until *before* or *after* the tail call?

Reply:
> What if the old function has return type `bool`, but the callee has return type `u8` and returns `2` -- is that UB? I would think so but currently I think that would be missed. 

I haven't tested it yet, but I actually think this won't be missed, because the check is in the `eval_fn_call`: https://github.com/rust-lang/rust/blob/af3f2f84d34cd8397e2ce36d91541acd38b1e223/compiler/rustc_const_eval/src/interpret/terminator.rs#L578-L587

But `pop_stack_frame` indeed does a lot more (like deallocating locals). @RalfJung what do you think would be the best to implement necessary things for tail calls here? Just copy all the necessary bits into `eval_terminator` (in the `TailCall` arm)? Should we try sharing similar bits between `pop_stack_frame` and tail calls (that potentially seems annoying...)? Should we introduce a new hook (`M::before_tail_call`) (should it default to `M::after_stack_pop`?)?

> In particular this is relevant for the aliasing model, where protectors are active until the stack frame pops.
>
> This is actually an interesting question for the aliasing model... should protector remain active _until_ before or _after_ the tail call?

This is indeed an interesting question... I'd assume protectors remain active until _before_ the tail call, since we basically replace the current stack frame with a new one.

Reply:
> I haven't tested it yet, but I actually think this won't be missed, because the check is in the eval_fn_call: 

That check will treat bool and u8 as being compatible, so it won't catch this problem.

Given that the caller basically delegates its return place to the callee, I think it would be reasonable to say that the caller's type no longer matters. However this means we cannot easily turn a tail call into a regular call as the validity requirements may change then.

> I'd assume protectors remain active until before the tail call, since we basically replace the current stack frame with a new one.

This depends on whether we allow the compiler to turn tail calls into regular calls, assuming it can know that this won't lead to exploding stack usage.

I'm okay with punting on the aliasing question for now, we should just make sure to track it somewhere.

More interesting is the unwinding question. If a nounwind function tailcalls to a mayunwind function, and then unwinds, that must still be UB. However that might already be taken care of? The new stack frame will inherit the `return_to_block` info from the old one, in particular the unwind action, so if that says `Unreachable` we'll still trigger UB as desired. I assume a tail call itself doesn't have an unwind action, it just gets inherited from the current stack frame?

> what do you think would be the best to implement necessary things for tail calls here? Just copy all the necessary bits into eval_terminator (in the TailCall arm)? Should we try sharing similar bits between pop_stack_frame and tail calls (that potentially seems annoying...)? Should we introduce a new hook (M::before_tail_call) (should it default to M::after_stack_pop?)?

I think we want a `replace_stack_frame` function in eval_context.rs, so all this code is reasonably close to each other. And if there are significant parts that are shared with push_stack_frame and pop_stack_frame, that should probably move into helper functions.

Reply:
> Given that the caller basically delegates its return place to the callee, I think it would be reasonable to say that the caller's type no longer matters. However this means we cannot easily turn a tail call into a regular call as the validity requirements may change then.

By "caller's type no longer matters" do you mean that in this scenario `c` is sound? (`b` definitely isn't though)

```rust
fn a() -> u8 { 2 }

fn b() -> bool {
    become transmute::<fn() -> u8, fn() -> bool>(callee)();
}

fn c() {
    let _x: u8 = transmute::<fn() -> bool, fn() -> u8>(callee)()();
}
```

> I'm okay with punting on the aliasing question for now, we should just make sure to track it somewhere.

Can you open an issue to track this?

> More interesting is the unwinding question. If a nounwind function tailcalls to a mayunwind function, and then unwinds, that must still be UB. However that might already be taken care of? The new stack frame will inherit the `return_to_block` info from the old one, in particular the unwind action, so if that says `Unreachable` we'll still trigger UB as desired. I assume a tail call itself doesn't have an unwind action, it just gets inherited from the current stack frame?

Yes, this sounds correct to me.

> I think we want a `replace_stack_frame` function in eval_context.rs right besides push/pop_stack_frame, so all this code is reasonably close to each other. (Maybe we should move all this to a new stack.rs file or so.) And if there are significant parts that are shared with push_stack_frame and pop_stack_frame, that should probably move into helper functions.

Ok. I will look into implementing this

Reply:
Opened an issue: https://github.com/rust-lang/unsafe-code-guidelines/issues/491.